helix-term/src/ui/picker/query.rs

```diff
@@ -0,0 +1,191 @@
+use std::{collections::HashMap, sync::Arc};
+
+pub(super) type PickerQuery = HashMap<&'static str, Arc<str>>;
+
+pub(super) fn parse(
+    column_names: &[&'static str],
+    primary_column: usize,
+    input: &str,
+) -> PickerQuery {
+    let mut fields: HashMap<&'static str, String> = HashMap::new();
+    let primary_field = column_names[primary_column];
+    let mut escaped = false;
+    let mut quoted = false;
+    let mut in_field = false;
+    let mut field = None;
+    let mut text = String::new();
+
+    macro_rules! finish_field {
+        () => {
+            let key = field.take().unwrap_or(primary_field);
+
+            if let Some(pattern) = fields.get_mut(key) {
+                pattern.push(' ');
+                pattern.push_str(&text);
+                text.clear();
+            } else {
+                fields.insert(key, std::mem::take(&mut text));
+            }
+        };
+    }
+
+    for ch in input.chars() {
+        match ch {
+            // Backslash escaping
+            '\\' => escaped = !escaped,
+            _ if escaped => {
+                // Keep backslash escapes for special nucleo syntax.
+                if matches!(ch, '^' | '$' | '!' | '\'') {
+                    text.push('\\');
+                }
+                text.push(ch);
+                escaped = false;
+            }
+            // Double quoting
+            '"' => quoted = !quoted,
+            ' ' if quoted => {
+                text.push('\\');
```

Comment:
why the double backslash?

Reply:
Isn't that just how you have to escape a single backslash?

Reply:
Yeah ofcourse that is ovvious what I meant is why does a space get turned into `\ `

Reply:
IIRC it was trying to escape spaces so that `"a b.txt"` was equivalent to `a\ b.txt` but actually nucleo doesn't seem to care about the backslash. Is there a way to tell nucleo to respect a space?

Reply:
hmm it really should but seems you found a bug there. I will look into fixing that.

But I don't think the quoting should actually do this. The quoting IMO should only escape the column parsing but not the nucleo syntax (so quoted text should behave the same the picker does currently). There is a bunch of other nucleo syntax that this doesn't escape where it would be weird if just some of it is escaped and not all of it. I also specifically was looking for a quoted syntax like this because the AND search that words does is extremely useful (for example for matching paths which now require a colum prefix so I wanted to be able to do `%p:"helix-term lib.rs` for example.

Reply:
hmm I think that is your parser somehow actually. `\ ` works on master but not with this branch

Reply:
haven't look to closely at it yet

Reply:
I might just be mistaken and this block is out of place. I'll take a look and add some test cases for edge cases

Reply:
Yep I was consuming the escape for something else. Fixed in the latest